File: pyCmpl/lib/pyCmpl/CmplParameter.py
#***********************************************************************
 #  This code is part of pyCMPL 
 #
 #  Copyright (C) 2013
 #  Mike Steglich - Technical University of Applied Sciences
 #  Wildau, Germany 
 #
 #  pyCMPL is a project of the Technical University of 
 #  Applied Sciences Wildau and the Institute for Operations Research 
 #  and Business Management at the Martin Luther University 
 #  Halle-Wittenberg.
 #  Please visit the project homepage <www.coliop.org>
 # 
 #  pyCMPL is free software; you can redistribute it and/or modify it 
 #  under the terms of the GNU Lesser General Public License as published by 
 #  the Free Software Foundation; either version 3 of the License, or 
 #  (at your option) any later version.
 # 
 #  pyCMPL is distributed in the hope that it will be useful, but WITHOUT
 #  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 #  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public 
 #  License for more details.
 #  
 #  You should have received a copy of the GNU Lesser General Public License
 #  along with this program; if not, see <http://www.gnu.org/licenses/>.
 #
 #**********************************************************************

#!/usr/bin/python 

from __future__ import division
from math import *

from CmplException import *
from CmplTools import *
from CmplSet import *

		
#*************** CmplParameter ***********************************		
class CmplParameter(object):

	#*********** constructor *********
	def __init__(self, name, *sets ):
	
		self.__name = ""
		self.__rank = 0
		self.__valueList = []
		self.__count=1
		self.__defaultVal=None
		
		defSets=None
		
		if len(sets) == 0:
			defSets=sets
		else:
			if type(sets[0]) == list:
				defSets=sets[0]
			else:
				defSets=sets
	
		if type(name) != str:
			raise CmplException("not a valid name for a parameter: " + str(name) )
		elif CmplTools.stringContainsWhitespace(name):
			raise CmplException("parameter name " + str(name) + " contains whitespaces")
		else:
			self.__name = name
		
		self._setList = []
		self.__rank = 0
	
		for s in defSets:
			if type(s) != CmplSet:
				raise CmplException(str(s)+ " is not a CMPL set ")
			else: 
				self._setList.append(s)
				self.__rank+=s.rank
		
	#*********** End constructor ****
	
	#*********** valueCount *********	
	def __countElements(self, vList):
		_count=len(vList)
		if type(vList[0]) == list:
			_count *= self.__countElements(vList[0])
		return _count
	#*********** end valueCount *****
		
	
	#*********** values *************	
	def setValues(self, val , defaultVal=None):
		self.__valueList=[]
		if self.__rank == 0:
			if not 'LIST' in str(type(val)).upper():
			#if type(val) != list:
				self.__valueList.append(val)
			else:
				raise CmplException("incompatible data for a scalar parameter : " + str(val) )
		else:
			if  'LIST' in str(type(val)).upper():
			#if type(val) == list:
				self.__valueList=val
				sCount=1
				for s in self._setList:
					sCount*=s.count()
					
				self.__count = self.__countElements(val)		
				if self.__count !=  sCount :
					raise CmplException("The dimension of the paramter "+ self.__name +" doesn't match the dimension of the set(s)."  )

			elif  'DICT' in str(type(val)).upper():
			#elif type(val) == dict:
				self.__valueList=val
				
				# keys are to be checked here or not to be checked in here (then in CMPL)??

				if defaultVal!=None:
					#sparse array with default
					if not (type(defaultVal) != str or type(defaultVal) != int or type(defaultVal) != int or type(defaultVal) != float ):
						raise CmplException("Wrong type of the default value of the paramter "+ self.__name +" : " + str(defaultVal) + " : " + type(defaultVal)  )
						
					self.__defaultVal=defaultVal
				
				else:
					#dense array
					sCount=1
					sRank=0
					for s in self._setList:
						sCount*=s.count()
						
					if sCount!=len(val):
						raise CmplException("The dimension of the paramter "+ self.__name +" doesn't match the dimension of the set(s)."  )
					
					if 'LIST' in str(type(val.values()[0])).upper() or 'DICT' in str(type(val.values()[0])).upper():
					#if type(val.values()[0]) == list or type(val.values()[0])== dict:
						raise CmplException("Unexpected data type for paramter "+ self.__name   )
					
			else:
				raise CmplException("incompatible data for a parameter array: " + str(val) )
	
	#*********** end values *********
	
	#*********** setList ************	
	@property
	def setList(self) :
		return self._setList
	#*********** end setList ********
	
	#*********** valueList **********
	@property
	def values(self):
		return self.__valueList		
	#*********** End valueList ******	
		
	#*********** valueList **********
	@property
	def value(self):
		return self.__valueList[0]		
	#***	
		
	#*********** name ***************
	@property
	def name(self):
		return self.__name		
	#*********** end name ***********	
	
	#*********** rank ***************
	@property
	def rank(self):
		return self.__rank		
	#*********** end rank ***********
	
	#*********** len ***************
	@property
	def len(self):
		return self.__count		
	#*********** end rank ***********

	#*********** defaultValue ************	
	@property
	def defaultValue(self) :
		return self.__defaultVal
	#*********** end setList ********
		
#*************** end CmplParameter *******************************		
	
		
	

		
		
		
		

		

File: jCmpl/src/jCMPL/CmplParameter.java
/* ****************************************************************************
 * This code is part of jCMPL
 *
 * Copyright (C) 2013 Mike Steglich / B. Knie Technical University of Applied
 * Sciences Wildau, Germany
 *
 * jCMPL is a project of the Technical University of Applied Sciences Wildau
 * and the Institute for Operations Research and Business Management at the
 * Martin Luther University Halle-Wittenberg.
 *
 * Please visit the project homepage <http://www.coliop.org>
 *
 * jCMPL is free software; you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 3 of the License, or (at your option) any
 * later version.
 *
 * jCMPL is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Lesser General Public # License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 * ****************************************************************************/

package jCMPL;

import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.List;

/**
 * Handling of CMPL parameters in jCMPL
 */
public class CmplParameter 
{
    
    
    private String _name;
    private Object _valueList;
    private ArrayList<CmplSet> _setList;
    private long _count;
    private int _rank;
    private int _valType; // 1 List 2 Array 3 scalar
    
    /**
     * internal function for setting defaults
     * @param name Name of the set
     * @throws CmplException 
     */
    private void setDefaults(String name) throws CmplException {
        _count = 1;
        _rank = 0;

        Pattern pattern = Pattern.compile("\\s");
        Matcher matcher = pattern.matcher(name);

        if (!matcher.find()) {
            this._name = name;
        } else {
            throw new CmplException("parameter name " + _name + " contains whitespaces");
        }
        
        _rank = 0;
        _setList = new ArrayList<CmplSet>();
    }
    
    /**
     * Constructor for a scalar parameter 
     * @param name Name of the set
     * @throws CmplException 
     */
    public CmplParameter(String name) throws CmplException {
        setDefaults(name);
    }
    
    /**
     * Constructor for a parameter array
     * @param name Name of the set
     * @param sets CmplSets over that the parameter is defined
     * @throws CmplException 
     */
    public CmplParameter(String name, CmplSet... sets) throws CmplException {
        setDefaults(name);
        for (CmplSet set : sets) {
            _setList.add(set);
            _rank += set.rank();
        }
    }
    
    /**
     * Internal function to count all set elements
     * @throws CmplException 
     */
    private long countElements(Object val) throws CmplException {
        long _tmpcount = 0;
        if (val.getClass().toString().contains("List")) {
            
            _tmpcount = ((ArrayList)val).size();
            
            if (((ArrayList)val).get(0).getClass().toString().contains("List") || ((ArrayList)val).get(0).getClass().isArray() ) {
                 _tmpcount *= countElements( ((ArrayList)val).get(0));
            } 
        } else if (val.getClass().isArray()) {
            
              if (val instanceof int[]) {
                  _tmpcount = ((int[])val).length;
                  _tmpcount *= countElements( ((int[])val)[0] );
              } else if (val instanceof long[]) {
                  _tmpcount = ((long[])val).length;
                  _tmpcount *= countElements( ((long[])val)[0] );
              } else if (val instanceof float[]) {
                  _tmpcount = ((float[])val).length; 
                  _tmpcount *= countElements( ((float[])val)[0] );                  
              } else if (val instanceof double[]) {
                  _tmpcount = ((double[])val).length;
                   _tmpcount *= countElements( ((double[])val)[0] );
              } else if (val instanceof Object[]) {
                  _tmpcount = ((Object[])val).length;
                   _tmpcount *= countElements( ((Object[])val)[0] );
              } else {
                    throw new CmplException("parameter name " + _name + " containts data with unsupported data type" + val.getClass().toString());
              }
                
        } else { return 1; }
        
        return _tmpcount;
    }

    /**
     * Internal function to set the value of a scalar param
     * @param val value of the parameter
     * @throws CmplException 
     */
    private void setScalarParam(Object val) throws CmplException {
        if (_rank > 0) {
            throw new CmplException("incompatible data for a parameter array : " + val.toString());
        }

        _valueList = val;
        _valType = 3;
    }

    /**
     * Sets the value of a scalar Integer parameter
     * @param val value
     * @throws CmplException 
     */
    public void setValues(Integer val) throws CmplException {
        setScalarParam(val);
    }

    /**
     * Sets the value of a scalar Long parameter
     * @param val value
     * @throws CmplException
     */
    public void setValues(Long val) throws CmplException {
        setScalarParam(val);
    }
    
    
    /**
     * Sets the value of a scalar Float parameter
     * @param val value
     * @throws CmplException
     */
    public void setValues(Float val) throws CmplException {
        setScalarParam(val);
    }
    
     
   /**
     * Sets the value of a scaler Double parameter
     * @param val value
     * @throws CmplException
     */
    public void setValues(Double val) throws CmplException {
        setScalarParam(val);
    }
    

    /**
     * Sets the value of a scalar String parameter
     * @param val value
     * @throws CmplException
     */
    public void setValues(String val) throws CmplException {
        setScalarParam(val);
    }
    
    
  /**
   * Defines the values of a parameter array
   * @param val List or array of parameter values
   * @throws CmplException 
   */
    public void setValues(Object val) throws CmplException {
        
        if (!val.getClass().toString().contains("List") && !val.getClass().isArray()) {
            throw new CmplException("Incompatible data - not a list or an array - for parameter :" + _name);
        }
         
        if (_rank == 0) {
            throw new CmplException("incompatible data for a scalar parameter : " + val.toString());
        }

        _valueList =  val;
        
        if (val.getClass().toString().contains("List")) {
            _valType=1;
        } else if (val.getClass().isArray()) {
            _valType=2;
        }
       
        long sCount = 1;
        for (int i = 0; i < _setList.size(); i++) {
            sCount *= _setList.get(i).len();
        }

        _count = countElements(_valueList);
        if (_count != sCount) {
            throw new CmplException("The dimension of the paramter " + _name + " doesn't match the dimension of the set(s).");
        }
        
    }

    /**
     * Returns a List of the invoked sets
     * @return set list
     */
    public List<CmplSet> setList() {
        return _setList;
    }

    /**
     * List of the values of a parameter 
     * @return value list
     */
    public Object values() {
        return _valueList;
    }

    /**
     * Value of a scalar parameter 
     * @return value of the scalar parameter
     */
    public Object value() {
        return _valueList;
    }

    /**
     * Name of the parameter
     * @return  name of the CMPL parameter (not the name of the CmplParameter object)
     */
    public String name() {
        return _name;
    }

    /**
     * Rank of the parameter
     * @return rank of the CMPL parameter 
     */
    public int rank() {
        return _rank;
    }

    /**
     * Length of the parameter array
     * @return number of elements in the parameter array
     */
    public long len() {
        return _count;
    }
   
    /**
     * Returns whether the parameter is organized as list, array or single value
     * @return valType
     */
    public int valType() {
        return _valType;
    }
    
}
